frameioclient/download.py

```diff
@@ -1,18 +1,174 @@
+    try:
+      fp = open(self.destination, "wb")
+      fp.write(b'\0' * self.file_size)
+      fp.close()
+    except Exception as e:
+      print(e)
+      return False
+    return True
+
+  def get_download_key(self):
+    try:
+      url = self.asset['original']
+    except KeyError:
+      self.watermarked = True
+      if self.asset['is_session_watermarked'] == True:
+        for resolution_key, download_url in sorted(self.asset['downloads'].items()):
+          resolution = resolution_key.split("_")[1] # Grab the item at index 1 (resolution)
+          try:
+            resolution = int(resolution)
+            if download_url is not None:
+              self.resolution_map.update({
+                resolution: download_url
+              })
+          except ValueError:
+            continue
+
+        # Grab the highest resolution now
+        url = self.resolution_map.values()[0]
+      else:
+        raise WatermarkIDDownloadException
+
+    return url
+
+  def get_path(self):
+    if self.prefix != None:
+      self.filename = self.prefix + self.filename
+
+    if self.destination == None:
+      final_destination = os.path.join(self.download_folder, self.filename)
+      self.destination = final_destination
+      
+    return self.destination
+
+  def download_handler(self):
+    if os.path.isfile(self.get_path()):
+      print("File already exists at this location.")
+      return self.destination
+    else:
+      url = self.get_download_key()
+
+      if self.watermarked == True:
+        return self.download(url)
+      else:
+        if self.acceleration == True:
+          return self.accelerated_download(url)
+        else:
+          return self.download(url)
+
+  def download(self, url):
+    start_time = time.time()
+    print("Beginning download -- {} -- {}".format(self.asset['name'], format_bytes(self.file_size, type="size")))
+
+    # Downloading
     r = requests.get(url)
+    open(self.destination, 'wb').write(r.content)
+
+    download_time = time.time() - start_time
+    download_speed = format_bytes(math.ceil(self.file_size/(download_time)))
+    print("Downloaded {} at {}".format(self.file_size, download_speed))
+
+    return self.destination, download_speed
+
+  def accelerated_download(self, url):
+    start_time = time.time()
+
+    # Generate stub
+    try:
+      self._create_file_stub()
+
+    except Exception as e:
+      raise DownloadException
+      print("Aborting", e)
+
+    offset = math.ceil(self.file_size / self.chunks)
+    in_byte = 0 # Set initially here, but then override
+    
+    print("Accelerated download -- {} -- {}".format(self.asset['name'], format_bytes(self.file_size, type="size")))
+
+    # Queue up threads
+    with concurrent.futures.ThreadPoolExecutor(max_workers=self.concurrency) as executor:
+      for i in range(int(self.chunks)):
+        out_byte = offset * (i+1) # Advance by one byte to get proper offset
+        task = (url, in_byte, out_byte, i)
```

Comment:
Your chunk size has an additional byte  😜, I am not if this will work or why it is working

```
filesize = 25 bytes
chunks = 1
offset = math.ceil(self.file_size / self.chunks) = 25

index 0
out_byte = 25 * (0 + 1) = 25
Header Range: 0-25 = 26 bytes
```

My expectation would be
```
Header Range: 0-24 = 25 bytes = filesize
```

I had this issue too

Reply:
Oh interesting... I wonder if it's working simply because I'm asking for the extra byte each time and then overwriting it?